### PR TITLE
Add Server Deck tests

### PR DESCRIPTION
**Pull request by Qazzquimby**

### CODE REVIEW

File: Server/src/test/java/Deck_Test.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Test
+    public void hasCard_whenMissingGeneral_NotFindGeneral() {
+        Deck sut = new Deck("sut");
+        ServerCard general = makeGeneralCard();
+
+        boolean actualHasCard = sut.hasCard(general.getCardId());
+        boolean expectedHasCard = false;
+
+        Assert.assertEquals(expectedHasCard, actualHasCard);
```
Comment (fluffy-hamster) [inline Server/src/test/java/Deck_Test.java]: can simplify to assertTrue

File: Server/src/test/java/Deck_Test.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        sut.addCard(general0);
+        sut.addCard(general1);
+    }
+
+    @Test
+    public void removeCard_whenHasCard_removesCard() throws LogicException {
+        Deck sut = new Deck("sut");
+        ServerCard minion = makeMinionCard();
+        sut.addCard(minion);
+        sut.removeCard(minion);
```
Comment (fluffy-hamster) [inline Server/src/test/java/Deck_Test.java]: add assert after add to check card actually is added

File: Server/src/test/java/Deck_Test.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+        boolean actualHasCard = sut.hasCard(minion.getCardId());
+        boolean expectedHasCard = false;
+
+        Assert.assertEquals(expectedHasCard, actualHasCard);
+    }
+
+    @Test(expected = ClientException.class)
+    public void removeCard_whenMissinCard_throwClientError() throws LogicException {
```
Comment (fluffy-hamster) [inline Server/src/test/java/Deck_Test.java]: typo: missing

File: Server/src/test/java/Deck_Test.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        ServerCard minion = makeMinionCard();
+        sut.removeCard(minion);
+    }
+
+    @Test
+    public void isValid_fine_isTrue() throws LogicException {
+        Deck sut = new Deck("sut");
+        ServerCard general = makeGeneralCard();
+        sut.addCard(general);
+        for (int i = 0; i < 20; i++) {
```
Comment (fluffy-hamster) [inline Server/src/test/java/Deck_Test.java]: make 20 a variable

File: Server/src/test/java/Deck_Test.java (hunk 2 above)
Comment (Qazzquimby) [reply Server/src/test/java/Deck_Test.java]: I don't think so. Good to have one assertion per test, and checking that a card is added is already done.

**Comment by Qazzquimby** [comment]
I'm having trouble where gson keeps generating using "others" instead of "cards" even when there's no trace of "others" in the codebase. I'd think it had to do with maven not fully cleaning and building, but I've tried so many things.

File: Server/src/test/java/Deck_Test.java (hunk 2 above)
Comment (fluffy-hamster) [reply Server/src/test/java/Deck_Test.java]: As a general rule, fewer asserts is better.  If you have another test checking add then I'm fine omitting the 2nd assert.

File: Client/src/main/java/org/projectcardboard/client/models/card/Deck.java [hunk 5]
```diff
@@ -86,7 +86,7 @@ public int count(ICard card) {
             case MINION:
             case SPELL:
                 int count = 0;
-                for (Card other : others) {
+                for (Card other : cards) {
```
Comment (Ridleh) [inline Client/src/main/java/org/projectcardboard/client/models/card/Deck.java]: change other to card

File: Client/src/main/java/org/projectcardboard/client/models/card/Deck.java [hunk 6]
```diff
@@ -112,7 +112,7 @@ public boolean hasCard(Card other) {
     public Card getCard(String cardName) {
         if (hero != null && hero.isSameAs(cardName)) return hero;
 
-        for (Card other : others) {
+        for (Card other : cards) {
```
Comment (Ridleh) [inline Client/src/main/java/org/projectcardboard/client/models/card/Deck.java]: change other to card

File: Client/src/main/java/org/projectcardboard/client/models/card/ExportedDeck.java [hunk 7]
```diff
@@ -16,7 +16,7 @@ public ExportedDeck(Deck deck) {
         if (deck.getHero() != null) {
             heroName = deck.getHero().getName();
         }
-        for (Card other : deck.getOthers()) {
+        for (Card other : deck.getCards()) {
```
Comment (Ridleh) [inline Client/src/main/java/org/projectcardboard/client/models/card/ExportedDeck.java]: change variable name

File: Client/src/main/java/org/projectcardboard/client/models/card/TempDeck.java [hunk 8]
```diff
@@ -30,11 +29,7 @@ public String getHeroId() {
         return heroId;
     }
 
-    public String getItemId() {
-        return itemId;
-    }
-
-    public List<String> getOthersIds() {
```
Comment (Ridleh) [inline Client/src/main/java/org/projectcardboard/client/models/card/TempDeck.java]: Did you double check we arent using this method anywhere? sometimes IntelliJ reports that a method isnt being use when in fact it is (e.g. the setNewNextCard method on the client side)

File: Server/src/server/dataCenter/models/card/Deck.java [hunk 9]
```diff
@@ -93,47 +86,15 @@ public void removeCard(ServerCard card) throws ClientException {
...
-        List<ServerCard> oldOthers = this.others;
-        this.others = new ArrayList<>();
-        for (ServerCard other : oldOthers) {
-            ServerCard card = new ServerCard(other);
-            card.setCardId(makeId(card, numberOf(card.getName()) + 1));
-            others.add(card);
-        }
-    }
-
-    private String makeId(ServerCard card, int number) {
```
Comment (Ridleh) [inline Server/src/server/dataCenter/models/card/Deck.java]: if this method isnt used, where else are we creating card IDs?

File: Server/src/server/dataCenter/models/card/ExportedDeck.java [hunk 10]
```diff
@@ -9,18 +9,14 @@
...
     public ExportedDeck(Deck deck) {
         name = deck.getName();
         if (deck.getHero() != null) {
             heroName = deck.getHero().getName();
         }
-        if (deck.getItem() != null) {
-            itemName = deck.getItem().getName();
-        }
-        for (Card other : deck.getOthers()) {
+        for (Card other : deck.getCards()) {
```
Comment (fluffy-hamster) [inline Server/src/server/dataCenter/models/card/ExportedDeck.java]: other

File: Client/src/main/java/org/projectcardboard/client/models/card/TempDeck.java (hunk 8 above)
Comment (Qazzquimby) [reply Client/src/main/java/org/projectcardboard/client/models/card/TempDeck.java]: getItemId? No instances in the project.

File: Server/src/server/dataCenter/models/card/Deck.java (hunk 9 above)
Comment (Qazzquimby) [reply Server/src/server/dataCenter/models/card/Deck.java]: Server Collection.addCard and a shared Card constructor. I'm not sure where the best place is but short term I think we're better without unused code.

**Comment by Qazzquimby** [comment]
I think everything's dealt with

**Pull request merged**